Mathlib/Combinatorics/SimpleGraph/Operations.lean

```diff
@@ -29,30 +30,125 @@ section ReplaceVertex
+  ext e; refine' e.inductionOn _
+  simp only [replaceVertex, mem_edgeSet, Set.mem_union, Set.mem_diff, mk'_mem_incidenceSet_iff]
+  intros; split_ifs; exacts [by simp_all, by aesop, by rw [adj_comm]; aesop, by aesop]
+
+variable [Fintype V] [DecidableRel G.Adj]
+
+instance : DecidableRel (G.replaceVertex s t).Adj := by unfold replaceVertex; infer_instance
+
+theorem edgeFinset_replaceVertex_of_not_adj (hn : ¬G.Adj s t) : (G.replaceVertex s t).edgeFinset =
+    G.edgeFinset \ G.incidenceFinset t ∪ (G.neighborFinset s).image (s(·, t)) := by
+  simp only [incidenceFinset, neighborFinset, ← Set.toFinset_diff, ← Set.toFinset_image,
+    ← Set.toFinset_union]
+  exact Set.toFinset_congr (G.edgeSet_replaceVertex_of_not_adj hn)
+
+theorem edgeFinset_replaceVertex_of_adj (ha : G.Adj s t) : (G.replaceVertex s t).edgeFinset =
+    (G.edgeFinset \ G.incidenceFinset t ∪ (G.neighborFinset s).image (s(·, t))) \ {s(t, t)} := by
+  simp only [incidenceFinset, neighborFinset, ← Set.toFinset_diff, ← Set.toFinset_image,
+    ← Set.toFinset_union, ← Set.toFinset_singleton]
+  exact Set.toFinset_congr (G.edgeSet_replaceVertex_of_adj ha)
+
+lemma disjoint_sdiff_neighborFinset_image :
+    Disjoint (G.edgeFinset \ G.incidenceFinset t) ((G.neighborFinset s).image (s(·, t))) := by
+  rw [disjoint_iff_ne]
+  intro e he
+  have : t ∉ e := by
+    rw [mem_sdiff, mem_incidenceFinset] at he
+    obtain ⟨_, h⟩ := he
+    contrapose! h
+    simp_all [incidenceSet]
+  aesop
+
+theorem card_edgeFinset_replaceVertex_of_not_adj (hn : ¬G.Adj s t) :
+    (G.replaceVertex s t).edgeFinset.card = G.edgeFinset.card + G.degree s - G.degree t := by
+  have inc : G.incidenceFinset t ⊆ G.edgeFinset := by simp [incidenceFinset, incidenceSet_subset]
+  rw [G.edgeFinset_replaceVertex_of_not_adj hn,
+    card_disjoint_union G.disjoint_sdiff_neighborFinset_image, card_sdiff inc,
+    tsub_add_eq_add_tsub <| card_le_card inc, card_incidenceFinset_eq_degree]
+  congr 2
+  rw [card_image_of_injective, card_neighborFinset_eq_degree]
+  unfold Function.Injective
+  aesop
+
+theorem card_edgeFinset_replaceVertex_of_adj (ha : G.Adj s t) :
+    (G.replaceVertex s t).edgeFinset.card = G.edgeFinset.card + G.degree s - G.degree t - 1 := by
+  have inc : G.incidenceFinset t ⊆ G.edgeFinset := by simp [incidenceFinset, incidenceSet_subset]
+  rw [G.edgeFinset_replaceVertex_of_adj ha, card_sdiff (by simp [ha]),
+    card_disjoint_union G.disjoint_sdiff_neighborFinset_image, card_sdiff inc,
+    tsub_add_eq_add_tsub <| card_le_card inc, card_incidenceFinset_eq_degree]
+  congr 2
+  rw [card_image_of_injective, card_neighborFinset_eq_degree]
+  unfold Function.Injective
+  aesop
 
 end ReplaceVertex
+
+section AddEdge
+
+/-- Given a vertex pair, add the corresponding edge to the graph's edge set if not present. -/
+def addEdge : SimpleGraph V where
+  Adj v w := G.Adj v w ∨ s ≠ t ∧ (s = v ∧ t = w ∨ s = w ∧ t = v)
+  symm v w := by simp_rw [adj_comm]; (conv_lhs => arg 2; arg 2; rw [or_comm]); exact id
```

Comment:
I would prefer we use lattice operations for adding edges. The idea is to make constructor `SimpleGraph.edge (v w : V) : SimpleGraph V` that creates a graph with a single edge between `v` and `w` if they're not equal (and no edge if they are), and then `G.addEdge v w` would instead be `G ⊔ edge v w`. This is more versatile, though perhaps lemmas such as `addEdge_of_adj` are a bit more brittle to apply.

I won't hold up merging the PR on account of this; I don't feel like refactoring `addEdge` later would be very difficult.

Reply:
Well the refactor was indeed easy and is now at #9736.